pipeflow_lib.tlv

```diff
@@ -485,129 +472,113 @@ m4_popdef(['m4_out_stage'])
+            $empty = ! $two_valid && ! $valid_count[0];
+            $full = ($valid_count == full_mark_['']m4_plus_inst_id);  // Could optimize for power-of-two depth.
          /entry[*]
-            $is_head = /m4_top|m4_in_pipe/entry>>m4_align(m4_in_at + 1, m4_out_at)$is_head;
-            $pop  = $is_head && ! |m4_out_pipe$blocked;
-            /read_masked
-               m4_trans_hier
-m4_trans_ind               $ANY = /entry$is_head ? /m4_top|m4_in_pipe/entry['']m4_trans_hier>>m4_align(m4_in_at + 1, m4_out_at)$ANY /* $aNY */ : '0;
-            /accum
-               m4_trans_hier
-m4_trans_ind               $ANY = ((entry == 0) ? '0 : /entry[(entry+(m4_depth)-1)%(m4_depth)]/accum['']m4_trans_hier$ANY) |
-                          /entry/read_masked['']m4_trans_hier$ANY;
-         /head
-            $trans_avail = |m4_out_pipe$trans_avail;
-            ?$trans_avail
-               m4_trans_hier
-m4_trans_ind               $ANY = /m4_top|m4_out_pipe/entry[(m4_depth)-1]/accum['']m4_trans_hier$ANY;
-
-   // Bypass
-   |m4_out_pipe
-      @m4_out_at
-         // Available output.  Sometimes it's necessary to know what would be coming to determined
-         // if it's blocked.  This can be used externally in that case.
-         /fifo_head
-            $trans_avail = |m4_out_pipe$trans_avail;
-            ?$trans_avail
-               m4_trans_hier
-m4_trans_ind               $ANY = /m4_top|m4_in_pipe>>m4_reverse_bypass_align$would_bypass
-m4_trans_ind                            ? /m4_top|m4_in_pipe['']m4_trans_hier>>m4_reverse_bypass_align$ANY
-m4_trans_ind                            : |m4_out_pipe/head['']m4_trans_hier$ANY;
-         $trans_avail = ! /m4_top|m4_in_pipe>>m4_reverse_bypass_align$would_bypass || /m4_top|m4_in_pipe>>m4_reverse_bypass_align$trans_avail;
-         $trans_valid = $trans_avail && ! $blocked;
-         ?$trans_valid
-            m4_trans_hier
-m4_trans_ind            $ANY = |m4_out_pipe/fifo_head['']m4_trans_hier$ANY;
-
+            @m4_eval(@m4_in_at + 1)
+               $prev_entry_state = /entry[(entry+(#m4_depth)-1)%(#m4_depth)]$state;
+               $next_entry_state = /entry[(entry+1)%(#m4_depth)]$state;
+               $reconstructed_is_tail = (  /m4_top|m4_in_pipe$two_valid && (!$state && $prev_entry_state)) ||
+                                        (! /m4_top|m4_in_pipe$two_valid && (!$next_entry_state && $state)) ||
+                                        (|m4_in_pipe$empty && (entry == 0));  // need a tail when empty for push
+               $is_head = $state && ! $prev_entry_state;
+               $reconstructed_valid = $state || (/m4_top|m4_in_pipe$two_valid && $prev_entry_state);
+      // Write data
+      |m4_in_pipe
+         @m4_in_at
+            /entry[*]
+               //?$push
+               //   $aNY = |m4_in_pipe['']m4_trans_hier$ANY;
+               #m4_trans_hier
+   m4_trans_ind            $ANY = /entry$push ? /m4_top|m4_in_pipe['']#m4_trans_hier$ANY : >>1$ANY /* RETAIN */;
```

Comment:
Maybe these lines are the reason you indented? The indentation is meaningful in TLV and will impact the design. m4_trans_ind is a string of spaces and is used for parameterized indentation. You can add a space in front and delete a space after. I have to make a change to sandpiper to address this properly.

Reply:
ok  now i realize it 
the reason is also m4_trans_ind  didn't make makerchip view the colored indentation like it used to so that's the reason of the mistake

Reply:
ok